Referensi: https://en.wikipedia.org/wiki/Random_forest

## General ML Model Training

Proses training selalu diawali dengan "Training set",di dalamnya terdiri dari sekumpulan training features "X_train" dan juga sekumpulan target label "Y_train".training set digunakan untuk melakukan training model machine learning dan model yang sudah di training akan dikenal dengan istilah "trained model".trained model ini akan kita gunakan untuk melakukan prediksi terhadap sekumpulan nilai features yang baru "X_new" dan prediksi yang dihasilkan oleh trained model dikenal dengan istilah "y_pred"

![title](img12.png)

## Ensemble Learning: heterogeneous & homogeneous

Referensi: https://en.wikipedia.org/wiki/Ensemble_learning

Ensemble Learning -> untuk menggabungkan beberapa model untuk melakukan prediksi

Ensemble learning juga diawali dari suatu training set.hanya saja kali ini training set yang tersedia akan kita gunakan untuk mentraining beberapa model,terdapat tiga model machine learning yang akan kita training yaitu KNN,SVM,dan Decision tree.

Proses training ini menghasilkan tiga model :  Trained model 1,Trained model 2,Trained model 3.setiap trained model ini akan kita gunakan untuk melakukan prediksi terhadap sekumpulan nilai features yang baru.prediksi yang dihasilkan setiap trained model direpresentasikan sebagai y_pred 1,y_pred 2,y_pred 3.karena terdapat lebih dari satu prediksi,maka perlu disatukan dengan suatu mekanisme yang dikenal dengan istilah "Majority Voting".pada kasus regression task majority voting dilakukan dengan menggunakan nilai mean atau nilai rata-rata sedangkan pada kasus classification task majority voting dicapai dengan menerapkan mode atau kemunculan terbanyak-banyak.ilustrasi hasil prediksi final direpresentasikan sebagai "y_pred final"

Karena disini kita memadukan sejumlah model machine learning yang berbeda maka seringkali dikenal juga sebagai "heterogeneous ensemble learning" sedangkan untuk kasus dimana kita memadukan model machine learning yang sejenis akan dikenal dengan istilah "homogeneous ensemble learning"

![title](img13.png)

## Bagging: Bootstrap Aggregating

Referensi: https://en.wikipedia.org/wiki/Bootstrap_aggregating

Kita akan mempelajari homogeneous ensemble learning dimana kita akan memadukan sejumlah model machine learning yang sejenis.

Proses ini juga diawali dengan training set,hanya saja disini karena kita menerapkan sejumlah model yang sejenis, maka melakukan training terhadap setiap model menggunakan training set yang sama bisa dikatakan sia-sia.karena akan menghasilkan trained model yang sama persis.disini kita akan berkenalan dengan suatu teknik yang dikenal dengan istilah bootstrap aggregating (bagging)

Mekanisme bagging pada dasarnya menerapkan proses random sampling with replacement terhadap training set yang kita miliki dan akan menghasilkan beberapa training set baru sejumlah model yang akan kita training.Training set baru yang dihasilkan oleh proses bagging ini dikenal dengan istilah "bag".karena kita akan mentraining tiga buah model,maka kita juga perlu persiapkan tiga buah bag :bag 1,bag 2,bag 3.ketiga bag ini berisi sekumpulan sample data yang berbeda satu dengan lainnya.karena ketiga bag ini dihasilkan dari proses random sampling with replacement yang diambil dari training set.

Setiap bag ini kita gunakan untuk mentraining tiga model machine learning yang sejenis.setelah melalui proses training.maka akan menghasilkan trained model (tiga trained model).setiap trained model ini berikutnya akan kita gunakan untuk melakukan prediksi terhadap sekumpulan nilai features yang baru.hasil prediksi yang dihasilkan setiap model selanjutnya akan kita satukan melalui proses majority voting untuk dihasilkan prediksi final.ketiga buah hasil prediksi tersebut kita satukan untuk menghasilkan hasil prediksi final

![title](img14.png)

## Random Forest

Referensi: https://en.wikipedia.org/wiki/Random_forest

Random forest merupakan implementasi dari homogeneous ensemble learning yang menerapkan decision tree,terbentuk dari sekumpulan dari decision tree.selain bagging,random forest juga menerapkan fitur randomness,dimana untuk setiap bag yang dihasilkan akan mengadopsi sejumlah features yang dipilih secara acak dari training set sumbernya atau dengan kata lain setiap model decision tree akan di training dengan bag yang berisi dataset yang beragam,tidak hanya dalam hal baris data aja,melainkan juga beragam dalam hal feature yang dihasilkan.karena tiap decision tree di training dengan menggunakan dataset yang berbeda,maka dihasilkanlah sejumlah trained model yang beragam.walau kesemua trained ini sebenarnya menggunakan model machine learning yang sejenis yaitu decision tree.setiap trained model ini selanjutnya akan kita gunakan untuk melakukan prediksi terhadap sekumpulan nilai features yang baru.prediksi yang dihasilkan oleh setiap trained model akan disatukan melalui proses majority voting untuk menghasilkan nilai prediksi final.

Strategi semacam ini menjadikan random forest memiliki performa yang jauh lebih superior bila dibandingkan model decision tree yang biasa.karena setiap tree yang ada di trained menggunakan subset dari keseluruhan training set yang tersedia.

![title](img15.png)

## Dataset

Referensi: https://en.wikipedia.org/wiki/Iris_flower_data_set

In [1]:
from sklearn.datasets import load_iris #import module

X, y = load_iris(return_X_y=True)

print(f'Dimensi Feature: {X.shape}')
print(f'Class: {set(y)}')

Dimensi Feature: (150, 4)
Class: {0, 1, 2}


* Hasil dimensi features adalah (150,4) yang artinya terdapat 150 baris dan 4 kolom
* Hasil class adalah {0,1,2} yang artinya terdapat tiga kelas

Training dan testing dataset tersebut dengan cara berikut

In [ ]:
from sklearn.model_selection import train_test_split #import module

X_train, X_test, y_train, y_test = train_test_split(X,
                                                    y,
                                                    test_size=0.3, #testing set akan menempati 30% dari total keseluruhan dataset
                                                    random_state=0) #menjamin agar eksperimen kita ini bisa kita replicate dan menghasilkan nilai yang konsisten

## Classification dengan RandomForestClassifier

In [ ]:
from sklearn.ensemble import RandomForestClassifier #import module

model = RandomForestClassifier(n_estimators=100, #berapa banyak decision tree yang ingin kita training
                               random_state=0)

model.fit(X_train, y_train)

## Evaluasi Model

In [ ]:
from sklearn.metrics import classification_report #import module

y_pred = model.predict(X_test)

print(classification_report(y_test, y_pred))